### PR TITLE
Move database role to url parameters from connection properties on Snowflake

### LINKED ISSUE #43600: Database role is not respected when private key is used in Snowflake
As per discussion in https://github.com/metabase/metabase/issues/41852 database role is not respected as connection property when there is connection url containing private key file.

However, if the role is set as url parameter, it is used normally. Hence we should set the role as url parameter.

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes https://github.com/metabase/metabase/issues/43600.

Role connection property is not respected when url with private key is present. This PR moves the role from connection properties to url params if url is used.

### CODE REVIEW

File: modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj [hunk 1]
```diff
@@ -422,7 +422,14 @@
...
                               :private-key-value (u/encode-base64 pk-key)
                               :user pk-user))]
-       (is (driver/can-connect? :snowflake details))))))
+       (testing "Can connect with base64 encoded `:private-key-value` and no `:private-key-options` set (#41852)"
+         (is (driver/can-connect? :snowflake details)))
+       ;; Following is required when private key and role are used together. See
+       ;; the [[metabase.driver.snowflake/maybe-add-role-to-spec-url]] for the details.
+       (testing "Role is added to connection url, if url is present (#43600)"
+         (is (str/includes?
+              (sql-jdbc.conn/connection-details->spec :snowflake (assoc details :role "SOME_ROLE"))
```
Comment (metamben) [inline modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:431]: Should we take the `:connection-uri` here?

**Review by metamben — APPROVED** [review]
Approving with some questions. I'm quite surprised the tests pass, I thought we were calling `str/includes?` on a map.

File: modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj (hunk 1 above)
Comment (lbrdnk) [reply modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj]: Yes, thanks for catching that. Reason the test was passing is that internally `str/includes?` call `.toStirng` first.

File: modules/drivers/snowflake/src/metabase/driver/snowflake.clj [hunk 2]
```diff
@@ -125,6 +125,22 @@
...
     (str "\"" (str/replace raw-name "\"" "\"\"") "\"")))
 
+(defn- maybe-add-role-to-spec-url
+  "Maybe add role to `spec`'s `:connection-uri`. This is necessary for rsa auth to work, because at the time of writing
+  Snowflake jdbc driver ignores `:role` connection property when `:connection-uri` (presumably containing
+  private_key_file) is used."
+  [spec details]
+  (if (and (string? (not-empty (:connection-uri spec)))
+           (string? (not-empty (:role details)))
+           (not (str/includes? (:connection-uri spec) "role=")))
```
[the remark this answers is not in the document]
Comment (lbrdnk) [reply modules/drivers/snowflake/src/metabase/driver/snowflake.clj:135]: I believe `role=` can occur only as a parameter of well formed uri string.

**Comment by lbrdnk** [comment]
@metabase-bot backport release-x.49.x

**Pull request merged**